tower/src/multiplex/service.rs

```diff
@@ -0,0 +1,100 @@
+use super::{Pick, Picker};
+use crate::util::Either;
+use crate::BoxError;
+use futures_util::ready;
+use pin_project::pin_project;
+use std::task::{Context, Poll};
+use std::{future::Future, pin::Pin};
+use tower_service::Service;
+
+/// TODO(david): docs
+#[derive(Debug, Clone, Copy)]
+pub struct Multiplex<A, B, P> {
+    picker: P,
+    first: A,
+    first_ready: bool,
+    second: B,
+    second_ready: bool,
+}
+
+impl<A, B, P> Multiplex<A, B, P> {
+    /// TODO(david): docs
+    pub fn new(first: A, second: B, picker: P) -> Self {
+        Self {
+            picker,
+            first,
+            first_ready: false,
+            second,
+            second_ready: false,
+        }
+    }
+}
+
+impl<R, P, A, B> Service<R> for Multiplex<A, B, P>
+where
+    P: Picker<R>,
+    A: Service<R>,
+    A::Error: Into<BoxError>,
+    B: Service<R, Response = A::Response>,
+    B::Error: Into<BoxError>,
+{
+    type Response = A::Response;
+    type Error = BoxError;
+    type Future = ResponseFuture<A::Future, B::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
```

Comment:
I feel like this style of polling will be limited for a general implementation. Tonic and Warp can get away because we can guarantee that our server services generally don't need to be polled ready (they are always ready).

Reply:
In my experience it hasn't been an issue since the services you combine are at the leaves and generally always are ready. But it is something to be aware of.

I'm not sure there is a nice way around it 🤔